File: galite-core/src/main/java/org/kopi/vkopi/lib/ui/swing/form/DField.java
/*
 * Copyright (c) 1990-2021 kopiRight Managed Solutions GmbH
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License version 2.1 as published by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 * $Id: DField.java 34997 2016-12-01 09:51:43Z hacheni $
 */

package org.kopi.vkopi.lib.ui.swing.form;

import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.event.ActionEvent;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;

import javax.swing.AbstractAction;
import javax.swing.ImageIcon;
import javax.swing.JButton;
import javax.swing.JPanel;

import org.kopi.galite.base.UComponent;
import org.kopi.galite.form.UBlock;
import org.kopi.galite.form.UField;
import org.kopi.galite.form.VBlock;
import org.kopi.galite.form.VConstants;
import org.kopi.galite.form.VField;
import org.kopi.galite.form.VFieldUI;
import org.kopi.galite.form.VForm;
import org.kopi.vkopi.lib.ui.swing.base.FieldStates;
import org.kopi.vkopi.lib.ui.swing.base.JFieldButton;
import org.kopi.vkopi.lib.ui.swing.base.Utils;
import org.kopi.vkopi.lib.ui.swing.visual.SwingThreadHandler;
import org.kopi.galite.visual.Action;
import org.kopi.galite.visual.VColor;

/**
 * DField is a panel composed in a text field and an information panel
 * The text field appear as a JLabel until it is edited
 */
public abstract class DField extends JPanel implements UField {

    // ----------------------------------------------------------------------
    // CONSTRUCTION
    // ----------------------------------------------------------------------

    public DField(VFieldUI model,
                  DLabel label,
                  int align,
                  int options,
                  boolean detail) {
        setLayout(new BorderLayout());
        addMouseListener(new DFieldMouseListener());

        this.inDetail = detail;
        this.model = model;
        this.options = options;
        this.label = label;
        this.align = align;
        isEditable = (options & VConstants.FDO_NOEDIT) == 0;

        if ((!getModel().getBlock().isMulti() || getModel().getBlock().noChart() || isInDetail())) {
            //&& (getModel().getDefaultAccess() >= VConstants.ACS_SKIPPED)) { //!!! hidden fields can be visible in other block modes.
            JPanel optionPane = new JPanel();

            optionPane.setLayout(new BorderLayout());
            if (model.hasAutofill()) {
                info = new JFieldButton(listImg);
                info.addActionListener(new AbstractAction() {
                    /**
                     * Comment for <code>serialVersionUID</code>
                     */
                    private static final long serialVersionUID = 2172836625163999381L;

                    public void actionPerformed(ActionEvent e) {
                        getModel().getForm().performAsyncAction(new Action("autofill") {
                            public void execute() {
                                DField.this.model.transferFocus(DField.this);
                                DField.this.model.autofillButton();
                            }
                        });
                    }
                });
                info.setEnabled(getModel().getDefaultAccess() > VConstants.ACS_SKIPPED);

                optionPane.add(info, BorderLayout.WEST);
            }
            if (model.getDecrementCommand() != null) {
                decr = new JFieldButton(leftImg);
                decr.addActionListener(new AbstractAction() {
                    /**
                     * Comment for <code>serialVersionUID</code>
                     */
                    private static final long serialVersionUID = -6114904737254808842L;

                    public void actionPerformed(ActionEvent e) {
                        DField.this.model.getDecrementCommand().performAction();
                    }
                });
                optionPane.add(decr, BorderLayout.CENTER);
            }
            if (model.getIncrementCommand() != null) {
                incr = new JFieldButton(rightImg);
                incr.addActionListener(new AbstractAction() {
                    /**
                     * Comment for <code>serialVersionUID</code>
                     */
                    private static final long serialVersionUID = -3670989616287237005L;

                    public void actionPerformed(ActionEvent e) {
                        DField.this.model.getIncrementCommand().performAction();
                    }
                });
                optionPane.add(incr, BorderLayout.EAST);
            }
            add(optionPane, BorderLayout.EAST);
        }
    }


    /**
     * Returns the object associed to record r
     *
     * @return the displayed value at this position
     */
    public abstract Object getObject();


    /**
     * Field cell renderer
     */
    public void setPosition(int pos) {
        this.pos = pos;
    }

    /**
     * Field cell renderer
     *
     * @return the position in chart (0..nbDisplay)
     */
    public int getPosition() {
        return pos;
    }

    /**
     * Returns the alignment
     */
    public int getAlign() {
        return align;
    }

    // ----------------------------------------------------------------------
    // UI MANAGEMENT
    // ----------------------------------------------------------------------

    /*package*/ void enter(boolean refresh) {
        updateFocus();
    }

    /*package*/ void leave() {
        updateFocus();
    }

    // ----------------------------------------------------------------------
    // DRAWING
    // ----------------------------------------------------------------------

    public void updateText() {
    }

    public void updateFocus() {
        SwingThreadHandler.verifyRunsInEventThread("DField.updateFocus");
        if (modelHasFocus()) {
            final VForm form = getModel().getForm();

            form.setInformationText(getModel().getToolTip());
            form.setFieldSearchOperator(getModel().getSearchOperator());
        }
    }

    public void forceFocus() {
        // not used
    }

    public void updateAccess() {
        SwingThreadHandler.verifyRunsInEventThread("DField.updateAccess");
        access = getAccess();

        setVisible(access != VConstants.ACS_HIDDEN);
        fireMouseHasChanged(); // $$$

        update(info);
        update(incr);
        update(decr);
    }

    /**
     * This method is called after an action of the user, object should
     * be redisplayed accordingly to changes.
     */
    public void update() {
        // overridden in subclasses
    }

    /**
     *
     */
    public UBlock getBlockView() {
        return model.getBlockView();
    }

    protected final boolean modelHasFocus() {
        if (getModel() == null) {
            return false;
        }

        final VBlock block = getModel().getBlock();
        return getModel().hasFocus()
                && block.getActiveRecord() == model.getBlockView().getRecordFromDisplayLine(pos);
    }

    protected final boolean isSkipped() {
        final VBlock block = getModel().getBlock();
        return getAccess() == VConstants.ACS_SKIPPED
                || !block.isRecordAccessible(model.getBlockView().getRecordFromDisplayLine(getPosition()));
    }

    public final int getAccess() {
        return getAccessAt(getPosition());
    }

    public final Color getForeground() {
        return getForegroundAt(getPosition());
    }

    public final Color getBackground() {
        return getBackgroundAt(getPosition());
    }

    protected final int getAccessAt(int at) {
        if (getModel() != null) {
            return getModel().getAccess(model.getBlockView().getRecordFromDisplayLine(at));
        } else {
            return VConstants.ACS_SKIPPED;
        }
    }

    protected final Color getForegroundAt(int at) {
        if (model != null) {
            VColor foreground;

            foreground = getModel().getForeground(model.getBlockView().getRecordFromDisplayLine(at));
            return foreground == null ? null : new Color(foreground.getRGB());
        } else {
            return null;
        }
    }

    protected final Color getBackgroundAt(int at) {
        if (model != null) {
            VColor background;

            background = getModel().getBackground(model.getBlockView().getRecordFromDisplayLine(at));
            return background == null ? null : new Color(background.getRGB());
        } else {
            return null;
        }
    }

    public final VField getModel() {
        return model.getModel();
    }

    public final void fireMouseHasChanged() {
        final int localAccess = getAccess();
        final int oldState = state;

        state = 0;

        if (modelHasFocus()) {
            state |= FieldStates.FOCUSED;
        }
        if (!isEditable) {
            state |= FieldStates.NOEDIT;
        }
        if (getModel().getBorder() == VConstants.BRD_HIDDEN) {
            state |= FieldStates.NOBORDER;
        }

        VBlock block = model.getBlock();

        if (block.isMulti() && !block.noChart() && !isInDetail()) {
            state |= FieldStates.CHART;
            if (block.getActiveRecord() == model.getBlockView().getRecordFromDisplayLine(pos)) {
                state |= FieldStates.ACTIVE;
            }
        } else {
            state |= FieldStates.ACTIVE;
        }

        if (mouseInside) {
            state |= FieldStates.ROLLOVER;
        }

        switch (localAccess) {
            case VConstants.ACS_HIDDEN:
                state |= FieldStates.HIDDEN;
                break;
            case VConstants.ACS_MUSTFILL:
                state |= FieldStates.MUSTFILL;
                break;
            case VConstants.ACS_VISIT:
                state |= FieldStates.VISIT;
                break;
            default:
                state |= FieldStates.SKIPPED;
                break;
        }

        if (oldState != state) {
            setDisplayProperties();
        }
    }

    public UComponent getAutofillButton() {
        return info;
    }

    public VFieldUI getRowController() {
        return model;
    }

    protected abstract void setDisplayProperties();

    // ----------------------------------------------------------------------
    // SNAPSHOT PRINTING
    // ----------------------------------------------------------------------

    /**
     * prepare a snapshot
     *
     * @param fieldPos position of this field within block visible fields
     */
    public void prepareSnapshot(int fieldPos, boolean activ) {
        label.prepareSnapshot(activ);
        if (activ) {
            state |= FieldStates.FOCUSED;
            setDisplayProperties();
        } else {
            state |= FieldStates.SKIPPED;
            setDisplayProperties();
        }

        if (info != null) {
            info.setVisible(false);
        }
        if (incr != null) {
            incr.setVisible(false);
        }
        if (decr != null) {
            decr.setVisible(false);
        }
    }

    // ----------------------------------------------------------------------
    // IMPLEMENTATION
    // ----------------------------------------------------------------------

    private void update(final JButton button) {
        if (button != null) {
            boolean was = button.isEnabled();
            boolean will = access >= VConstants.ACS_VISIT;
            if (was != will) {
                button.setEnabled(will);
            }
        }
    }

    public void setInDetail(boolean detail) {
        inDetail = detail;
        label.setInDetailMode(detail);
    }

    public boolean isInDetail() {
        return inDetail;
    }

    // ----------------------------------------------------------------------
    // INNER CLASSES
    // ----------------------------------------------------------------------

    protected final class DFieldMouseListener extends MouseAdapter {
        public void mouseClicked(MouseEvent e) {
            performAction();
        }

        public void mouseEntered(MouseEvent e) {
            mouseInside = true;
            fireMouseHasChanged();
        }

        public void mouseExited(MouseEvent e) {
            mouseInside = false;
            fireMouseHasChanged();
        }

        private void performAction() {
            if (!model.hasAction() && !modelHasFocus()) {
                // an empty row in a chart has not calculated
                // the access for each field (ACCESS Trigger)
                if (model.getBlock().isMulti()) {
                    int recno = model.getBlockView().getRecordFromDisplayLine(DField.this.getPosition());

                    if (!model.getBlock().isRecordFilled(recno)) {
                        model.getBlock().updateAccess(recno);
                    }
                }

                if (!model.getBlock().isMulti()
                        || model.getBlock().isDetailMode() == isInDetail()
                        || model.getBlock().noChart()) {
                    Action action = new Action("mouse1") {
                        public void execute() {
                            model.transferFocus(DField.this); // use here a mouse transferfocus
                        }
                    };
                    // execute it as model transforming thread
                    // it is not allowed to execute it not with
                    // the method performAsync/BasicAction.
                    model.performAsyncAction(action);
                }
            }
        }
    }

    // ----------------------------------------------------------------------
    // DATA MEMBERS
    // ----------------------------------------------------------------------

    protected VFieldUI model;
    public DLabel label;
    protected JFieldButton info;
    protected JFieldButton incr;
    protected JFieldButton decr;

    protected int state;        // Display state
    protected int pos;
    protected int options;
    protected int align;
    protected int access;        // current access of field
    protected boolean isEditable;    // is this field editable
    protected boolean mouseInside;    // private events

    private boolean inDetail;

    private static final ImageIcon listImg = Utils.getImage("list.gif");
    private static final ImageIcon rightImg = Utils.getImage("arrowright.gif");
    private static final ImageIcon leftImg = Utils.getImage("arrowleft.gif");
}

File: galite-core/src/main/kotlin/org/kopi/galite/visual/ui/vaadin/form/DField.kt
/*
 * Copyright (c) 2013-2022 kopiLeft Services SARL, Tunis TN
 * Copyright (c) 1990-2022 kopiRight Managed Solutions GmbH, Wien AT
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License version 2.1 as published by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */
package org.kopi.galite.visual.ui.vaadin.form

import org.kopi.galite.visual.base.UComponent
import org.kopi.galite.visual.form.UBlock
import org.kopi.galite.visual.form.UField
import org.kopi.galite.visual.form.VConstants
import org.kopi.galite.visual.form.VField
import org.kopi.galite.visual.form.VFieldUI
import org.kopi.galite.visual.form.VForm
import org.kopi.galite.visual.form.VImageField
import org.kopi.galite.visual.form.VStringField
import org.kopi.galite.visual.ui.vaadin.base.BackgroundThreadHandler.access
import org.kopi.galite.visual.ui.vaadin.base.StyleManager
import org.kopi.galite.visual.ui.vaadin.base.Styles
import org.kopi.galite.visual.ui.vaadin.field.AbstractField
import org.kopi.galite.visual.ui.vaadin.field.Field
import org.kopi.galite.visual.ui.vaadin.field.FieldListener
import org.kopi.galite.visual.ui.vaadin.visual.VApplication
import org.kopi.galite.visual.Action
import org.kopi.galite.visual.ApplicationContext
import org.kopi.galite.visual.VColor

import com.vaadin.flow.component.AttachEvent
import com.vaadin.flow.component.UI

/**
 * The `DField` is the vaadin [UField] implementation.
 *
 * @param model The field model.
 * @param label The field label.
 * @param align The field alignment.
 * @param options The field options.
 * @param inDetail Is it a detail view ?
 */
abstract class DField(internal var model: VFieldUI,
                      internal var label: DLabel?,
                      align: Int,
                      protected var options: Int,
                      private var inDetail: Boolean)
  : Field(model.getIncrementCommand() != null,
          model.getDecrementCommand() != null), UField, FieldListener {

  // ----------------------------------------------------------------------
  // DATA MEMBERS
  // ----------------------------------------------------------------------
  protected var state = 0 // Display state
  protected var pos = 0

  /** The alignment. */
  var align: Int = align
    protected set
  internal var access = 0 // current access of field
  protected var isEditable = options and VConstants.FDO_NOEDIT == 0 // is this field editable
  protected var mouseInside = false // private events
  protected val styleManager: StyleManager by lazy {
    (ApplicationContext.applicationContext.getApplication() as VApplication).styleManager
  }

  /**
   * The visible field height needed to create layout.
   */
  val visibleHeight: Int

  init {
    className = Styles.FIELD
    visibleHeight = when {
      getModel() is VStringField -> {
        (getModel() as VStringField).getVisibleHeight()
      }
      getModel() is VImageField -> {
        /*
         * Sets the visible height of the image field to allow row span in simple block layouts.
         * We estimate that a row height is ~ 20 px
         */
        (getModel() as VImageField).iconHeight / 20
      }
      else -> {
        getModel().height
      }
    }
    label!!.hasAction = model.hasAction()
  }

  fun setFieldContent(component: AbstractField<*>) {
    wrappedField = component
    wrappedField.addClickListener {
      // no click event is for rich text field and action fields
      /*if (hasAction || content is RichTextField) { TODO
        return
      }*/
      (model.blockView as DBlock).setActiveRecordFromDisplay(position) // TODO: do we need this?
      onClick()
    }
    add(component)
  }

  //----------------------------------------------------------------------
  // UI MANAGEMENT
  // ----------------------------------------------------------------------
  /**
   * Enters this field.
   * @param refresh Should we refresh GUI ?
   */
  open fun enter(refresh: Boolean) {
    updateFocus()
  }

  /**
   * Leaves this field.
   */
  open fun leave() {
    updateFocus()
  }

  //-------------------------------------------------
  // ACCESSORS
  //-------------------------------------------------
  /**
   * Field cell renderer
   * @return the position in chart (0..nbDisplay)
   */
  override var position: Int = 0

  final override fun getModel(): VField {
    return model.model
  }

  override fun setInDetail(detail: Boolean) {
    inDetail = detail
    label!!.isInDetail = detail
  }

  /**
   * Returns `true` is the field belongs to the detail view.
   * @return `true` is the field belongs to the detail view.
   */
  fun isInDetail(): Boolean {
    return inDetail
  }

  override fun getAutofillButton(): UComponent? = null

  /**
   * Returns the row controller.
   * @return The row controller.
   */
  val rowController: VFieldUI
    get() = model
  //-------------------------------------------------
  // UTILS
  //-------------------------------------------------
  /**
   * This method is called after an action of the user, object should
   * be redisplayed accordingly to changes.
   */
  open fun update() {
    // overridden in subclasses
  }

  override fun updateText() {
    // overridden in subclasses
  }

  override fun updateFocus() {
    if (modelHasFocus()) {
      val form: VForm = getModel().getForm()
      form.setInformationText(getModel().toolTip)
      form.setFieldSearchOperator(getModel().getSearchOperator())
    }
  }

  override fun forceFocus() {
    // to be implemented by subclasses
  }

  override fun updateAccess() {
    access(currentUI) {
      access = getAccess()
      updateStyles(access)
      isVisible = access != VConstants.ACS_HIDDEN
      update(label)
    }
  }


  var currentUI: UI? = null

  override fun onAttach(attachEvent: AttachEvent) {
    currentUI = attachEvent.ui
  }

  /**
   * Updates a given field label.
   * @param label The label to be updated.
   */
  private fun update(label: DLabel?) {
    if (label != null) {
      val was = label.isEnabled
      val will = access >= VConstants.ACS_VISIT
      if (was != will) {
        label.isEnabled = will
      }
    }
  }

  /**
   * Update field style according to its access.
   * @param access The field access.
   */
  private fun updateStyles(access: Int) {
    classNames.remove("visit")
    classNames.remove("skipped")
    classNames.remove("mustfill")
    classNames.remove("hidden")
    when (access) {
      VConstants.ACS_VISIT -> classNames.add("visit")
      VConstants.ACS_SKIPPED -> classNames.add("skipped")
      VConstants.ACS_MUSTFILL -> classNames.add("mustfill")
      VConstants.ACS_HIDDEN -> classNames.add("hidden")
      else -> classNames.add("visit")
    }
  }

  //-------------------------------------------------
  // ABSTRACT METHODS
  //-------------------------------------------------
  /**
   * Returns the object associated to current record
   *
   * @return        the displayed value at current position
   */
  abstract override fun getObject(): Any?

  /**
   * set blink state
   * @param blink The blink ability.
   */
  abstract override fun setBlink(blink: Boolean)

  abstract fun valueChanged()

  //-------------------------------------------------
  // PROTECTED UTILS
  //-------------------------------------------------
  /**
   * Returns `true` if the field model is focused.
   * @return `true` if the field model is focused.
   */
  protected fun modelHasFocus(): Boolean {
    val block = getModel().block
    return getModel().hasFocus() && block!!.activeRecord == getBlockView().getRecordFromDisplayLine(position)
  }

  /**
   * Returns `true` if the field is in skipped mode.
   * @return `true` if the field is in skipped mode.
   */
  protected fun isSkipped(): Boolean {
    val block = getModel().block
    return (getAccess() == VConstants.ACS_SKIPPED
            || !block!!.isRecordAccessible(getBlockView().getRecordFromDisplayLine(position)))
  }

  override fun getAccess(): Int {
    return getAccessAt(position)
  }

  /**
   * Returns the field access at a given line.
   * @param at The desired line.
   * @return The field access.
   */
  protected fun getAccessAt(at: Int): Int {
    return getModel().getAccess(getBlockView().getRecordFromDisplayLine(at))
  }

  /**
   * Returns the field foreground color.
   * @param at The desired line.
   * @return The foreground color.
   */
  protected fun getForegroundAt(at: Int): VColor? {
    return getModel().getForeground(getBlockView().getRecordFromDisplayLine(at))
  }

  /**
   * Returns the field background color.
   * @param at The desired line.
   * @return The background color.
   */
  protected fun getBackgroundAt(at: Int): VColor? {
    return getModel().getBackground(getBlockView().getRecordFromDisplayLine(at))
  }

  /**
   * Returns the foreground color of the current data position.
   * @return The foreground color of the current data position.
   */
  val foreground: VColor?
    get() = getForegroundAt(position)

  /**
   * Returns the background color of the current data position.
   * @return The background color of the current data position.
   */
  val background: VColor?
    get() = getBackgroundAt(position)
  // ----------------------------------------------------------------------
  // SNAPSHOT PRINTING
  // ----------------------------------------------------------------------
  /**
   * prepare a snapshot
   * @param fieldPos position of this field within block visible fields
   */
  override fun prepareSnapshot(fieldPos: Int, activ: Boolean) {
    label!!.prepareSnapshot(activ)
  }

  //-------------------------------------------------
  // FIELD IMPLEMENTATION
  //-------------------------------------------------
  override fun getBlockView(): UBlock {
    return model.blockView
  }

  override fun isEnabled(): Boolean { return super.isEnabled() }

  override fun setEnabled(enabled: Boolean) { super.setEnabled(enabled) }

  //-------------------------------------------------
  // FIELD LISTENER IMPLEMENTATION
  //-------------------------------------------------
  override fun onIncrement() {
    model.getIncrementCommand()!!.performAction()
  }

  override fun onDecrement() {
    model.getDecrementCommand()!!.performAction()
  }

  override fun onClick() {
    if (!modelHasFocus()) {
      // an empty row in a chart has not calculated
      // the access for each field (ACCESS Trigger)
      if (model.getBlock().isMulti()) {
        val recno: Int = getBlockView().getRecordFromDisplayLine(position)
        if (!model.getBlock().isRecordFilled(recno)) {
          model.getBlock().updateAccess(recno)
        }
      }
      if (!model.getBlock().isMulti()
              || model.getBlock().isDetailMode == isInDetail() || model.getBlock().noChart()) {
        val action: Action = object : Action("mouse1") {
          override fun execute() {
            model.transferFocus(this@DField) // use here a mouse transferfocus
          }
        }
        // execute it as model transforming thread
        // it is not allowed to execute it not with
        // the method performAsync/BasicAction.
        model.performAsyncAction(action)
      }
    }
  }

  override fun fireAction() {
    model.executeAction()
  }

  /**
   * !!! We never change transfer a focus to a field that belongs
   * to another block than this field model block. If we do it, it
   * can cause assertion errors when validating blocks caused by
   * actors actions.
   */
  override fun transferFocus() {
    if (!modelHasFocus()) {
      // an empty row in a chart has not calculated
      // the access for each field (ACCESS Trigger)
      if (model.getBlock().isMulti()) {
        val recno: Int = getBlockView().getRecordFromDisplayLine(position)
        if (!model.getBlock().isRecordFilled(recno)) {
          model.getBlock().updateAccess(recno)
        }
      }
      if (!model.getBlock().isMulti()
              || model.getBlock().isDetailMode == isInDetail() || model.getBlock().noChart()) {
        val action: Action = object : Action("mouse1") {
          override fun execute() {
            // proceed only of we are in the same block context.
            if (getModel().block == getModel().getForm().getActiveBlock()) {
              val recno: Int = getBlockView().getRecordFromDisplayLine(position)

              // go to the correct record if necessary
              // but only if we are in the correct block now
              if (getModel().block!!.isMulti()
                      && recno != getModel().block!!.activeRecord
                      && getModel().block!!.isRecordAccessible(recno)) {
                getModel().block!!.gotoRecord(recno)
              }

              // go to the correct field if already necessary
              // but only if we are in the correct record now
              if (recno == getModel().block!!.activeRecord
                      && getModel() != getModel().block!!.activeField
                      && getAccess() >= VConstants.ACS_VISIT) {
                getModel().block!!.gotoField(getModel())
              }
            }
          }
        }
        // execute it as model transforming thread
        // it is not allowed to execute it not with
        // the method performAsync/BasicAction.
        model.performAsyncAction(action)
      }
    }
  }

  /**
   * Tells the server side the the action field should be performed
   */
  fun actionPerformed() {
    if (model.hasAction() && access >= VConstants.ACS_VISIT) {
      fireAction()
    }
  }

  //---------------------------------------------------
  // NAVIGATION
  //---------------------------------------------------

  override fun gotoNextField() {
    getModel().getForm().performAsyncAction(object : Action("keyKEY_TAB") {
      override fun execute() {
        getModel().block!!.form.getActiveBlock()!!.gotoNextField()
      }
    })
  }

  override fun gotoPrevField() {
    getModel().getForm().performAsyncAction(object : Action("keyKEY_STAB") {
      override fun execute() {
        getModel().block!!.form.getActiveBlock()!!.gotoPrevField()
      }
    })
  }

  override fun gotoNextEmptyMustfill() {
    getModel().getForm().performAsyncAction(object : Action("keyKEY_ALTENTER") {
      override fun execute() {
        getModel().block!!.form.getActiveBlock()!!.gotoNextEmptyMustfill()
      }
    })
  }

  override fun gotoPrevRecord() {
    getModel().getForm().performAsyncAction(object : Action("keyKEY_REC_UP") {
      override fun execute() {
        getModel().block!!.gotoPrevRecord()
      }
    })
  }

  override fun gotoNextRecord() {
    getModel().getForm().performAsyncAction(object : Action("keyKEY_REC_DOWN") {
      override fun execute() {
        getModel().block!!.gotoNextRecord()
      }
    })
  }

  override fun gotoFirstRecord() {
    getModel().getForm().performAsyncAction(object : Action("keyKEY_REC_FIRST") {
      override fun execute() {
        getModel().block!!.form.getActiveBlock()!!.gotoFirstRecord()
      }
    })
  }

  override fun gotoLastRecord() {
    getModel().getForm().performAsyncAction(object : Action("keyKEY_REC_LAST") {
      override fun execute() {
        getModel().block!!.form.getActiveBlock()!!.gotoLastRecord()
      }
    })
  }

  /**
   * Performs the auto fill action.
   */
  fun performAutoFillAction() {
    getModel().getForm().performAsyncAction(object : Action("autofill") {
      override fun execute() {
        model.transferFocus(this@DField)
        model.autofillButton()
      }
    })
  }

  /**
   * Performs the field action trigger
   */
  fun performFieldAction() {
    if (model.hasAction()) {
      getModel().getForm().performAsyncAction(object : Action("TRG_ACTION") {
        override fun execute() {
          model.transferFocus(this@DField)
          getModel().callTrigger(VConstants.TRG_ACTION)
        }
      })
    }
  }
}
